python/src/iceberg/expression/base.py

```diff
@@ -0,0 +1,146 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from abc import ABC, abstractmethod
+from functools import reduce
+
+from iceberg.types import Singleton
+
+
+class BooleanExpression(ABC):
+    """base class for all boolean expressions"""
+
+    @abstractmethod
+    def __invert__(self) -> "BooleanExpression":
+        ...
+
+
+class And(BooleanExpression):
+    """AND operation expression - logical conjunction"""
+
+    def __new__(cls, left: BooleanExpression, right: BooleanExpression, *rest: BooleanExpression):
+        if rest:
+            return reduce(And, (left, right, *rest))
+        if left is alwaysFalse() or right is alwaysFalse():
+            return alwaysFalse()
+        elif left is alwaysTrue():
+            return right
+        elif right is alwaysTrue():
+            return left
+        return super().__new__(cls)
+
+    def __init__(self, left: BooleanExpression, right: BooleanExpression, *rest: BooleanExpression):
```

Comment:
Should this ensure that `rest` is empty since it should only be handled using `__new__`? If someone calls this directly, it would currently discard `rest`.

Reply:
Sorry, I was wrong. If `rest` is non-empty then it ignores `left` and `right`.

Reply:
If rest is non empty then new uses `reduce` to construct the object with only `left` and `right` specified

Reply:
Yes, but not in this method. If this method is called directly then it would do the wrong thing.